### PR TITLE
Uniqueness of first column is now checked before flagging it as a primary key

### PR DESCRIPTION
**Pull request by simha104**
### Pull Request Description
``IDColumnsDataCheck`` now only returns an action code to set the first column as the primary key if it contains unique values

### CODE REVIEW

File: evalml/tests/data_checks_tests/test_id_columns_data_check.py
Comment (Cmancuso) [inline evalml/tests/data_checks_tests/test_id_columns_data_check.py:355]: How is this working exactly? How can we tell which columns the recommendation is to drop and which columns it recommends to mark primary?

Comment (simha104) [reply evalml/tests/data_checks_tests/test_id_columns_data_check.py:355]: It is checking if the first column satisfies these two characteristics:

- It is either named 'ID' or its name ends with '_id'
- All of it's values are unique

If those characteristics are met, it suggests to mark as primary key. Otherwise, it suggests to drop columns

Comment (simha104) [reply evalml/tests/data_checks_tests/test_id_columns_data_check.py:355]: In this case , the columns name was 'ID' and had the values ["a", "b", "c", "d"] which are all unique

Comment (bchen1116) [reply evalml/tests/data_checks_tests/test_id_columns_data_check.py:355]: Right, I think the point of confusion comes in the metadata. When we look at the action code `SET_FIRST_COL_ID` and then look at the column metadata `['ID', 'col_2', ...]`, how do we know which of these is the col_id to set as the primary key? It would be clearer if the metadata had one value only when we need set the column so that we don't have to worry about confusion on later steps. The additional columns should be listed somewhere else.

Comment (simha104) [reply evalml/tests/data_checks_tests/test_id_columns_data_check.py:355]: Oh okay in the case where the data check action code is `SET_FIRST_COL_ID` , the first column name in the metadata is always the column to set as the primary key and the other columns are the ones to drop. Would it be a good idea for me to change the metadata structure to something like `{ "primary_key": "ID", "drop": ['col_2', 'col_3_id']`

Comment (bchen1116) [reply evalml/tests/data_checks_tests/test_id_columns_data_check.py:355]: Yea, this seems clearer to me! It makes more sense and will be much easier for developers down the line to understand

Comment (simha104) [reply evalml/tests/data_checks_tests/test_id_columns_data_check.py]: Okay done!

Comment (bchen1116) [reply evalml/tests/data_checks_tests/test_id_columns_data_check.py]: Ah, rather than using `drop` for the other columns, can you just continue to list them under `columns` value? I think that'd be fine rather than introducing a new value to check.

File: evalml/tests/data_checks_tests/test_id_columns_data_check.py
Comment (bchen1116) [inline evalml/tests/data_checks_tests/test_id_columns_data_check.py:230]: Curious what the others think about this. I think it'd be ideal if the `SET_FIRST_COL_ID` data check action is only associated with the column that we want to set as the primary key, and `DROP_COL` data check action to be associated with the other columns, rather than associating those columns with the `SET_FIRST_COL_ID` action. I don't think this is blocking, but might be something to file for future fix if we don't get to it now.

**Review by bchen1116 — COMMENTED** [review]
Left one non-blocking comment for discussion about how we do the data check actions, but the tests and implementation look good to me!

Would probably be good to update the [docs](https://feature-labs-inc-evalml--3639.com.readthedocs.build/en/3639/user_guide/data_checks.html#ID-Columns) to mention this additional capability of the ID column check!

File: docs/source/user_guide/data_checks.ipynb
Comment (bchen1116) [inline docs/source/user_guide/data_checks.ipynb:284]: Can we put something in here about how the primary key must is only recommended if first column is identified as ID?

File: docs/source/user_guide/data_checks.ipynb
Comment (bchen1116) [inline docs/source/user_guide/data_checks.ipynb:895]: Remove this

**Review by bchen1116 — COMMENTED** [review]
Great work on changing this! Left some nits and suggestions, but once those are addressed, lmk and I'll approve!

**Review by bchen1116 — APPROVED** [review]
Thanks for making the changes!

File: evalml/data_checks/id_columns_data_check.py
Comment (Cmancuso) [inline evalml/data_checks/id_columns_data_check.py:212]: would it make sense to just do the warning append in this conditional block to keep all that logic together?

File: evalml/data_checks/id_columns_data_check.py
Comment (Cmancuso) [inline evalml/data_checks/id_columns_data_check.py:92]: thanks for clearing this up, this text is much more helpful.

NAB but is this confusing? If the threshold is set to 1, I would sort of expect it to be identified as ID whether or not that was including in the column header. Might be worth discussing later.

File: evalml/data_checks/id_columns_data_check.py
Comment (simha104) [reply evalml/data_checks/id_columns_data_check.py]: Yeah you're right, made a change to remove the redundant logic

File: evalml/data_checks/id_columns_data_check.py
Comment (simha104) [reply evalml/data_checks/id_columns_data_check.py:92]: Should we reverse the threshold measurement such that the lower it is, the more selective `ID` column identification is?

**Review by Cmancuso — APPROVED** [review]
This is looking good to me, but lets make sure to get 1 review from the EvalML folks before merging.

File: docs/source/user_guide/data_checks.ipynb
Comment (eccabay) [inline docs/source/user_guide/data_checks.ipynb:284]: Can you add a bit more clarification for users here about what a primary key column is? There's useful information in here but having a clearer definition will make this more understandable!

File: evalml/data_checks/id_columns_data_check.py
Comment (eccabay) [inline evalml/data_checks/id_columns_data_check.py:126]: Nit: "is 100% likely" doesn't make much sense. Can you rephrase this to something along the lines of "has all unique values" instead?

File: evalml/data_checks/id_columns_data_check.py
Comment (eccabay) [inline evalml/data_checks/id_columns_data_check.py:211]: Nit: can we rephrase this to something like "The first column '{}' is likely to be the primary key"?

File: evalml/tests/data_checks_tests/test_id_columns_data_check.py
Comment (eccabay) [inline evalml/tests/data_checks_tests/test_id_columns_data_check.py:145]: This test has a lot of repeated code. You can clean this up by parameterizing over the data type and setting up the input data that way - see [here](https://github.com/alteryx/evalml/blob/main/evalml/tests/automl_tests/test_default_algorithm.py#L932) for an example

**Review by eccabay — APPROVED** [review]
I just left a couple small nitpicks, but otherwise looks good!

We're currently in the process of generating a release, so if you could hold off on merging this until after the release has gone out, that would be appreciated. It should be all set sometime today or early next week.

**Review by chukarsten — CHANGES_REQUESTED** [review]
Agreed with @eccabay , let's please use @pytest.mark.parametrize to vastly reduced repeated code in this PR.  Feel free to reach out to me if you need help with this!  Would be happy to show you how.

File: evalml/tests/conftest.py
Comment (chukarsten) [inline evalml/tests/conftest.py:94]: Clever, haven't seen a nested fixture like this.  Very cool.

**Review by chukarsten — APPROVED** [review]

**Pull request merged**
